Adding addtokenSigningcertificate (#11406)
microsoftgraph/microsoft-graph-docs: api-reference/beta/api/serviceprincipal-addtokensigningcertificate.md

---
title: "servicePrincipal: addTokenSigningCertificate"
description: "Add a signing certificate to a servicePrincipal."
localization_priority: Normal
author: "luleonpla"
ms.prod: "microsoft-identity-platform"
doc_type: "apiPageType"
---

> # servicePrincipal: addTokenSigningCertificate

Reviewer: Please add the new topics to the beta TOC.

Author: I have added the topic to the TOC.

---
title: "servicePrincipal: addTokenSigningCertificate"
description: "Add a signing certificate to a servicePrincipal."
localization_priority: Normal

> ms.author: "luleon"

Reviewer 2: ms.author isn't needed in the Graph repo.

author: "luleon"
ms.prod: "microsoft-identity-platform"
doc_type: "apiPageType"
---

# servicePrincipal: addKey

Namespace: microsoft.graph

[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]

> It creates a self-signed signing certificate and returns a [selfSignedCertificate](../resources/selfsignedcertificate.md) which is the public part of the generated certificate. The self-signed signing certificate is composed of 3 resources: the private key ([keyCredential](../resources/keycredential.md) with usage = 'Sign'), the public key ([keyCredential](../resources/keycredential.md) with usage = 'verify'), and the [passwordCredentials](../resources/passwordcredential.md).

Reviewer 2: ```suggestion
Creates a self-signed signing certificate and returns a [selfSignedCertificate](../resources/selfsignedcertificate.md), which is the public part of the generated certificate. The self-signed signing certificate is composed of these resources: the private key ([keyCredential](../resources/keycredential.md) with usage = 'Sign'), the public key ([keyCredential](../resources/keycredential.md) with usage = 'verify'), and the [passwordCredentials](../resources/passwordcredential.md).
```

> All 3 created resources have the same customKeyIdentifier.

Reviewer 2: ```suggestion
All of the created resources have the same customKeyIdentifier.
```

> The passwordCredential is used to open the pfx/private key. Also, it's associated with the privateKey having both the same `keyId`. 

Reviewer 2: These sentences do not need to be separate paragraphs.

The subject of the certificate is a constant value. It won't be affected by the optional displayName provided in the POST call.

The startDateTime is set to the same time the certificate is created using the action. 

The endDateTime can be up to 3 years after the certificate is created.

## Permissions

|Permission type      | Permissions (from least to most privileged)              |
|:--------------------|:---------------------------------------------------------|

> |Delegated (work or school account) | Application.ReadWrite.All, Directory.ReadWrite.All, Directory.AccessAsUser.All |

Reviewer 2: ```suggestion
|Delegated (work or school account) | Application.ReadWrite.All, Directory.ReadWrite.All |
```

|Delegated (personal Microsoft account) | None.    |
|Application | Application.ReadWrite.OwnedBy, Application.ReadWrite.All, Directory.ReadWrite.All |

## HTTP request

<!-- { "blockType": "ignored" } -->

```http
POST /servicePrincipals/{id}/addTokenSigningCertificate
```

## Request body

In the request body, provide the following required properties.

| Property	   | Type	|Description|
|:---------------|:--------|:----------|

> | Parameter             | Type     |Description|

Reviewer 2: Remove lines 53 and 54.

| --------------------- | -------- | --------- | ------------------------ |

> | `displayName` | `string` | Friendly name for the key.  It must start with 'CN='.|

Reviewer 2: ```suggestion
| displayName | string | Friendly name for the key.  It must start with `CN=`.|
```

> | `endDateTime` | `DateTimeOffset` |The date and time at which the credential expires. It can be up to 3 years from the date the certifcate is created. If not supplied, the default is 3 years from the time of creation. The Timestamp type represents date and time information using ISO 8601 format and is always in UTC time. For example, midnight UTC on Jan 1, 2014 would look like this: '2014-01-01T00:00:00Z' |

Reviewer 2: ```suggestion
| endDateTime | DateTimeOffset |The date and time when the credential expires. It can be up to 3 years from the date the certificate is created. If not supplied, the default is 3 years from the time of creation. The Timestamp type represents date and time information using ISO 8601 format and is always in UTC time. For example, midnight UTC on Jan 1, 2014 would look like this: '2014-01-01T00:00:00Z' .|
```

## Response

If successful, this method returns a `200 OK` response code and a new [selfSignedCertificate](../resources/selfsignedcertificate.md) object in the response body.

## Examples

### Request

The following is an example of the request.

> # [HTTP](#tab/http)

Reviewer 2: Remove line 68.

<!-- {
  "blockType": "request",
  "name": "serviceprincipal_addtokensigningcertificate"
}-->

```http

> POST https://graph.microsoft.com/beta/servicePrincipals/{id}/addTokenSigningCertificate

Reviewer 2: The example request should have a real looking id.
